core/include/moveit/task_constructor/stages/grasp_provider.h

```diff
@@ -0,0 +1,107 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Boston Cleek
+   Desc:   Grasp generator stage
+*/
+
+#pragma once
+
+#include <functional>
+#include <mutex>
+
+#include <moveit/task_constructor/stages/generate_pose.h>
+#include <moveit/task_constructor/stages/action_base.h>
+#include <moveit/task_constructor/storage.h>
+#include <moveit/task_constructor/marker_tools.h>
+#include <rviz_marker_tools/marker_creation.h>
+#include <moveit/planning_scene/planning_scene.h>
+
+namespace moveit {
+namespace task_constructor {
+namespace stages {
+
+/** @brief Generate grasp candidates using deep learning approaches */
+class GraspProvider : public GeneratePose, ActionBase
+{
+public:
+	/**
+	 * @brief Constructor
+	 * @param action_name - action namespace
+	 * @param stage_name - name of stage
+	 * @param goal_timeout - goal to completed time out (0 is considered infinite timeout)
+	 * @param server_timeout - connection to server time out (0 is considered infinite timeout)
+	 * @details Initialize the client and connect to server
+	 */
+	GraspProvider(const std::string& action_name, const std::string& stage_name = "grasp provider",
+	              double goal_timeout = 0.0, double server_timeout = 0.0);
+
+	/**
+	 * @brief Composes the action goal and sends to server
+	 */
+	void composeGoal();
+
+	/**
+	 * @brief Monitors status of action goal
+	 * @return true if grasp candidates are received within (optional) timeout
+	 * @details This is a blocking call. It will wait until either grasp candidates
+	 *          are received or the timeout has been reached.
+	 */
+	bool monitorGoal();
+
+	void activeCallback() override;
+	void feedbackCallback(const grasping_msgs::GraspPlanningFeedbackConstPtr& feedback) override;
+	void doneCallback(const actionlib::SimpleClientGoalState& state,
+	                  const grasping_msgs::GraspPlanningResultConstPtr& result) override;
+
+	void init(const core::RobotModelConstPtr& robot_model) override;
+	void compute() override;
+
+	void setEndEffector(const std::string& eef) { setProperty("eef", eef); }
+	void setObject(const std::string& object) { setProperty("object", object); }
+
+	void setPreGraspPose(const std::string& pregrasp) { properties().set("pregrasp", pregrasp); }
+	void setPreGraspPose(const moveit_msgs::RobotState& pregrasp) { properties().set("pregrasp", pregrasp); }
+	void setGraspPose(const std::string& grasp) { properties().set("grasp", grasp); }
+	void setGraspPose(const moveit_msgs::RobotState& grasp) { properties().set("grasp", grasp); }
+
+protected:
+	void onNewSolution(const SolutionBase& s) override;
+
+private:
+	std::mutex grasp_mutex_;  // Protects grasp candidates
+	std::atomic_bool found_candidates_;  // Flag indicates the discovery of grasps
+	std::vector<moveit_msgs::Grasp> grasp_candidates_;  // Grasp Candidates
```

Comment:
@JafarAbdi Can you take a look a these members? The intent is to address Robert's comment about thread safety and multiple feedback messages. I don't think the mutex in the case can cause deadlock. Please let me know if you see anything wrong with this.